### PR TITLE
sane-backends: 1.2.1 -> 1.3.1

### LINKED ISSUE #273280: USB scanner does not work because SANE tries to write to lock file in /nix/store
My Canon CanoScan LiDE 20 does not work on NixOS.

I wrote a minimal SANE config dir to reproduce:

```
$ find $PWD/sane-config-simple
/home/niklas/tmp/sane-config-simple
/home/niklas/tmp/sane-config-simple/plustek.conf
/home/niklas/tmp/sane-config-simple/dll.conf
```

```
$ cat sane-config-simple/dll.conf
plustek
```

```
$ cat sane-config-simple/plustek.conf                 
[usb] 0x04A9 0x220D
device auto
```

Invocation:

```sh
SANE_DEBUG_DLL=128 SANE_USB_WORKAROUND=1 SANE_DEBUG_SANEI_USB=128  SANE_DEBUG_PLUSTEK=255 SANE_CONFIG_DIR=$PWD/sane-config-simple strace -fye open,openat scanimage -L
````

Reveals:

<details>

```
[06:50:41.520033] [sanei_usb] libusb_scan_devices: skipping device 0x1d6b/0x0002 at 001:001: cannot open: Access denied (insufficient permissions)
[06:50:41.520094] [sanei_usb] sanei_usb_scan_devices: device 00 is libusb:001:015
[06:50:41.520153] [sanei_usb] sanei_usb_scan_devices: found 1 devices
[06:50:41.520216] [plustek] Plustek backend V0.52-12, part of sane-backends 1.2.1
[06:50:41.520280] [plustek] Retrieving all supported and conntected devices
[06:50:41.520356] [plustek] Checking for 0x07b3-0x0010
[06:50:41.520417] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0010
[06:50:41.520477] [plustek] Checking for 0x07b3-0x0011
[06:50:41.520537] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0011
[06:50:41.520596] [plustek] Checking for 0x07b3-0x0013
[06:50:41.520656] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0013
[06:50:41.520715] [plustek] Checking for 0x07b3-0x0017
[06:50:41.520774] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0017
[06:50:41.520834] [plustek] Checking for 0x07b3-0x0015
[06:50:41.520893] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0015
[06:50:41.520953] [plustek] Checking for 0x07b3-0x0015
[06:50:41.521012] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0015
[06:50:41.521071] [plustek] Checking for 0x07b3-0x0017
[06:50:41.521131] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0017
[06:50:41.521190] [plustek] Checking for 0x07b3-0x0013
[06:50:41.521250] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0013
[06:50:41.521309] [plustek] Checking for 0x07b3-0x0011
[06:50:41.521371] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0011
[06:50:41.521431] [plustek] Checking for 0x07b3-0x0010
[06:50:41.521491] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0010
[06:50:41.521551] [plustek] Checking for 0x07b3-0x0014
[06:50:41.521612] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0014
[06:50:41.521672] [plustek] Checking for 0x07b3-0x0014
[06:50:41.521731] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0014
[06:50:41.521791] [plustek] Checking for 0x07b3-0x0016
[06:50:41.521851] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0016
[06:50:41.521911] [plustek] Checking for 0x07b3-0x0017
[06:50:41.521971] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0017
[06:50:41.522031] [plustek] Checking for 0x07b3-0x0017
[06:50:41.522090] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0017
[06:50:41.522150] [plustek] Checking for 0x07b3-0x0007
[06:50:41.522210] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0007
[06:50:41.522270] [plustek] Checking for 0x07b3-0x000f
[06:50:41.522329] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x000f
[06:50:41.522392] [plustek] Checking for 0x07b3-0x000f
[06:50:41.522455] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x000f
[06:50:41.522514] [plustek] Checking for 0x07b3-0x0005
[06:50:41.522601] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0005
[06:50:41.522662] [plustek] Checking for 0x07b3-0x0014
[06:50:41.522721] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0014
[06:50:41.522780] [plustek] Checking for 0x07b3-0x0012
[06:50:41.522839] [sanei_usb] sanei_usb_find_devices: vendor=0x07b3, product=0x0012
[06:50:41.522903] [plustek] Checking for 0x0400-0x1000
[06:50:41.522964] [sanei_usb] sanei_usb_find_devices: vendor=0x0400, product=0x1000
[06:50:41.523028] [plustek] Checking for 0x0400-0x1001
[06:50:41.523091] [sanei_usb] sanei_usb_find_devices: vendor=0x0400, product=0x1001
[06:50:41.523162] [plustek] Checking for 0x0400-0x1001
[06:50:41.523233] [sanei_usb] sanei_usb_find_devices: vendor=0x0400, product=0x1001
[06:50:41.523345] [plustek] Checking for 0x0458-0x2007
[06:50:41.523423] [sanei_usb] sanei_usb_find_devices: vendor=0x0458, product=0x2007
[06:50:41.523498] [plustek] Checking for 0x0458-0x2008
[06:50:41.523571] [sanei_usb] sanei_usb_find_devices: vendor=0x0458, product=0x2008
[06:50:41.523642] [plustek] Checking for 0x0458-0x2009
[06:50:41.523718] [sanei_usb] sanei_usb_find_devices: vendor=0x0458, product=0x2009
[06:50:41.523795] [plustek] Checking for 0x0458-0x2013
[06:50:41.523869] [sanei_usb] sanei_usb_find_devices: vendor=0x0458, product=0x2013
[06:50:41.523945] [plustek] Checking for 0x0458-0x2015
[06:50:41.524021] [sanei_usb] sanei_usb_find_devices: vendor=0x0458, product=0x2015
[06:50:41.524096] [plustek] Checking for 0x0458-0x2016
[06:50:41.524169] [sanei_usb] sanei_usb_find_devices: vendor=0x0458, product=0x2016
[06:50:41.524326] [plustek] Checking for 0x03f0-0x0505
[06:50:41.524407] [sanei_usb] sanei_usb_find_devices: vendor=0x03f0, product=0x0505
[06:50:41.524479] [plustek] Checking for 0x03f0-0x0605
[06:50:41.524550] [sanei_usb] sanei_usb_find_devices: vendor=0x03f0, product=0x0605
[06:50:41.524619] [plustek] Checking for 0x04b8-0x010f
[06:50:41.524688] [sanei_usb] sanei_usb_find_devices: vendor=0x04b8, product=0x010f
[06:50:41.524761] [plustek] Checking for 0x04b8-0x011d
[06:50:41.524832] [sanei_usb] sanei_usb_find_devices: vendor=0x04b8, product=0x011d
[06:50:41.524906] [plustek] Checking for 0x1606-0x0050
[06:50:41.524977] [sanei_usb] sanei_usb_find_devices: vendor=0x1606, product=0x0050
[06:50:41.525049] [plustek] Checking for 0x1606-0x0060
[06:50:41.525118] [sanei_usb] sanei_usb_find_devices: vendor=0x1606, product=0x0060
[06:50:41.525187] [plustek] Checking for 0x1606-0x0160
[06:50:41.525258] [sanei_usb] sanei_usb_find_devices: vendor=0x1606, product=0x0160
[06:50:41.525330] [plustek] Checking for 0x049f-0x001a
[06:50:41.525409] [sanei_usb] sanei_usb_find_devices: vendor=0x049f, product=0x001a
[06:50:41.525479] [plustek] Checking for 0x04a9-0x2206
[06:50:41.525549] [sanei_usb] sanei_usb_find_devices: vendor=0x04a9, product=0x2206
[06:50:41.525644] [plustek] Checking for 0x04a9-0x2207
[06:50:41.525714] [sanei_usb] sanei_usb_find_devices: vendor=0x04a9, product=0x2207
[06:50:41.525780] [plustek] Checking for 0x04a9-0x2208
[06:50:41.525845] [sanei_usb] sanei_usb_find_devices: vendor=0x04a9, product=0x2208
[06:50:41.525912] [plustek] Checking for 0x04a9-0x220d
[06:50:41.525977] [sanei_usb] sanei_usb_find_devices: vendor=0x04a9, product=0x220d
[06:50:41.526045] [plustek] Checking for 0x04a9-0x220e
[06:50:41.526111] [sanei_usb] sanei_usb_find_devices: vendor=0x04a9, product=0x220e
[06:50:41.526177] [plustek] Checking for 0x04a9-0x2220
[06:50:41.526242] [sanei_usb] sanei_usb_find_devices: vendor=0x04a9, product=0x2220
[06:50:41.526306] [plustek] Checking for 0x0a82-0x4600
[06:50:41.526385] [sanei_usb] sanei_usb_find_devices: vendor=0x0a82, product=0x4600
[06:50:41.526451] [plustek] Checking for 0x0a82-0x6620
[06:50:41.526524] [sanei_usb] sanei_usb_find_devices: vendor=0x0a82, product=0x6620
[06:50:41.526590] [plustek] Checking for 0x04a7-0x0427
[06:50:41.526668] [sanei_usb] sanei_usb_find_devices: vendor=0x04a7, product=0x0427
[06:50:41.526727] [plustek] Checking for 0x0a53-0x1000
[06:50:41.526786] [sanei_usb] sanei_usb_find_devices: vendor=0x0a53, product=0x1000
[06:50:41.526845] [plustek] Checking for 0x0a53-0x2000
[06:50:41.526903] [sanei_usb] sanei_usb_find_devices: vendor=0x0a53, product=0x2000
[06:50:41.526962] [plustek] Available and supported devices:
[06:50:41.527021] [plustek] Device: >libusb:001:015< - 0x04a9x0x220d
[pid 63324] openat(AT_FDCWD</home/niklas/tmp>, "/home/niklas/tmp/sane-config-simple/plustek.conf", O_RDONLY) = 7</home/niklas/tmp/sane-config-simple/plustek.conf>
[06:50:41.527266] [plustek] ># Plustek-SANE Backend configuration file<
[06:50:41.527329] [plustek] ># For use with LM9831/2/3 based USB scanners<
[06:50:41.527397] [plustek] >#<
[06:50:41.527456] [plustek] ><
[06:50:41.527515] [plustek] ># each device needs at least two lines:<
[06:50:41.527573] [plustek] ># - [usb] vendor-ID and product-ID<
[06:50:41.527632] [plustek] ># - device devicename<
[06:50:41.527691] [plustek] ># i.e. for Plustek (0x07B3) UT12/16/24 (0x0017)<
[06:50:41.527749] [plustek] ># [usb] 0x07B3 0x0017<
[06:50:41.527808] [plustek] ># device /dev/usbscanner<
[06:50:41.527867] [plustek] ># or<
[06:50:41.527925] [plustek] ># device libusb:bbb:ddd<
[06:50:41.527984] [plustek] ># where bbb is the busnumber and ddd the device number<
[06:50:41.528043] [plustek] ># make sure that your user has access to /proc/bus/usb/bbb/ddd<
[06:50:41.528101] [plustek] >#<
[06:50:41.528160] [plustek] ># additionally you can specify some options<
[06:50:41.528218] [plustek] ># warmup, lOffOnEnd, lampOff<
[06:50:41.528277] [plustek] >#<
[06:50:41.528339] [plustek] ># For autodetection use<
[06:50:41.528398] [plustek] ># [usb]<
[06:50:41.528457] [plustek] ># device /dev/usbscanner<
[06:50:41.528515] [plustek] >#<
[06:50:41.528574] [plustek] ># or simply<
[06:50:41.528632] [plustek] ># [usb]<
[06:50:41.528691] [plustek] >#<
[06:50:41.528750] [plustek] ># or if you want a specific device but you have no idea about the<
[06:50:41.528808] [plustek] ># device node or you use libusb, simply set vendor- and product-ID<
[06:50:41.528867] [plustek] ># [usb] 0x07B3 0x0017<
[06:50:41.528926] [plustek] ># device auto<
[06:50:41.528985] [plustek] >#<
[06:50:41.529044] [plustek] ># NOTE: autodetection is safe, as it uses the info it got<
[06:50:41.529102] [plustek] >#       from the USB subsystem. If you're not using the<
[06:50:41.529161] [plustek] >#       autodetection, you MUST have attached that device<
[06:50:41.529220] [plustek] >#       at your USB-port, that you have specified...<
[06:50:41.529278] [plustek] >#<
[06:50:41.529628] [plustek] ><
[06:50:41.529713] [plustek] >[usb] 0x04A9 0x220D<
[06:50:41.529775] [plustek] next device is a USB device (0x04A9-0x220D)
[06:50:41.529834] [plustek] ... next device
[06:50:41.529893] [plustek] >#device libusb:001:015<
[06:50:41.529951] [plustek] >device auto<
[06:50:41.530010] [plustek] Decoding device name >auto<
[06:50:41.530097] [plustek] attach (auto, 0x7ffc5a863df0, (nil))
[06:50:41.530157] [plustek] Device configuration:
[06:50:41.530215] [plustek] device name  : >auto<
[06:50:41.530274] [plustek] USB-ID       : >0x04A9-0x220D<
[06:50:41.530332] [plustek] model ovr.   : 0
[06:50:41.530397] [plustek] warmup       : -1s
[06:50:41.530456] [plustek] lampOff      : -1
[06:50:41.530514] [plustek] lampOffOnEnd : yes
[06:50:41.530573] [plustek] cacheCalData : no
[06:50:41.530631] [plustek] altCalibrate : no
[06:50:41.530689] [plustek] skipCalibr.  : no
[06:50:41.530748] [plustek] skipFine     : no
[06:50:41.530806] [plustek] skipFineWhite: no
[06:50:41.530865] [plustek] skipDarkStrip: no
[06:50:41.530923] [plustek] incDarkTarget: yes
[06:50:41.530982] [plustek] invertNegs.  : no
[06:50:41.531040] [plustek] dis.Speedup  : no
[06:50:41.531098] [plustek] pos_x        : 0
[06:50:41.531157] [plustek] pos_y        : 0
[06:50:41.531215] [plustek] pos_shading_y: -1
[06:50:41.531273] [plustek] neg_x        : 0
[06:50:41.531332] [plustek] neg_y        : 0
[06:50:41.531394] [plustek] neg_shading_y: -1
[06:50:41.531453] [plustek] tpa_x        : 0
[06:50:41.531511] [plustek] tpa_y        : 0
[06:50:41.531570] [plustek] tpa_shading_y: -1
[06:50:41.531628] [plustek] red gain     : -1
[06:50:41.531686] [plustek] green gain   : -1
[06:50:41.531745] [plustek] blue gain    : -1
[06:50:41.531803] [plustek] red offset   : -1
[06:50:41.531862] [plustek] green offset : -1
[06:50:41.531921] [plustek] blue offset  : -1
[06:50:41.531979] [plustek] red lampoff  : -1
[06:50:41.532038] [plustek] green lampoff: -1
[06:50:41.532096] [plustek] blue lampoff : -1
[06:50:41.532155] [plustek] red Gamma    : 1.00
[06:50:41.532215] [plustek] green Gamma  : 1.00
[06:50:41.532274] [plustek] blue Gamma   : 1.00
[06:50:41.532338] [plustek] gray Gamma   : 1.00
[06:50:41.532398] [plustek] ---------------------
[06:50:41.532465] [plustek] usbDev_open(auto,0x04A9-0x220D) - 0x203f440
[pid 63324] openat(AT_FDCWD</home/niklas/tmp>, "/nix/store/4wxksva2132pc1ddy4rk83mx2j8q9p82-sane-backends-1.2.1/var/lock/LCK..libusb:001:015", O_WRONLY|O_CREAT|O_EXCL, 0644) = -1 EROFS (Read-only file system)
[06:50:41.532572] [plustek] sanei_access_lock failed: 11
[06:50:41.532633] [plustek] open failed: -1
[06:50:41.532691] [dll] init: backend `plustek' is version 1.0.0
[06:50:41.532750] [plustek] sane_get_devices (0x7ffc5a865f70, 0)
[06:50:41.532808] [dll] sane_get_devices: found 0 devices

No scanners were identified. If you were expecting something different,
check that the scanner is plugged in, turned on and detected by the
sane-find-scanner tool (if appropriate). Please read the documentation
which came with this software (README, FAQ, manpages).
[06:50:41.532953] [dll] sane_exit: exiting
[06:50:41.533015] [dll] sane_exit: calling backend `plustek's exit function
[06:50:41.533076] [plustek] sane_exit
[06:50:41.533210] [dll] sane_exit: finished
[pid 63325] +++ exited with 0 +++
+++ exited with 0 +++
```

</details>

Important part:

```c
[plustek] usbDev_open(auto,0x04A9-0x220D) - 0x203f440
openat(AT_FDCWD</home/niklas/tmp>, "/nix/store/4wxksva2132pc1ddy4rk83mx2j8q9p82-sane-backends-1.2.1/var/lock/LCK..libusb:001:015", O_WRONLY|O_CREAT|O_EXCL, 0644) = -1 EROFS (Read-only file system)
[plustek] sanei_access_lock failed: 11
[plustek] open failed: -1
```

It's failing because it's trying to write a lock file into `/nix/store`.

### PR DESCRIPTION
**Pull request by onny**
## Description of changes

Changelog: https://gitlab.com/sane-project/backends/-/releases/1.3.1

Fixes https://github.com/NixOS/nixpkgs/issues/273280

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by onny** [comment]
@gitwah please give it a try if the new version fixes the issue with your scanner :)

**Review by symphorien — CHANGES_REQUESTED** [review]
this regresses cross compilation

**Comment by onny** [comment]
Also looking to fix https://github.com/NixOS/nixpkgs/issues/273280

**Review by symphorien — APPROVED** [review]
the update looks sane. I did not test the locking change.

**Comment by gitwah** [comment]
> the update looks sane. I did not test the locking change.

The locking issue https://github.com/NixOS/nixpkgs/issues/273280 was fixed for my by this change.

File: pkgs/applications/graphics/sane/backends/default.nix [hunk 1]
```diff
@@ -1,57 +1,89 @@
...
-    (fetchpatch {
-      name = "compile-sane-desc-for-build.patch";
-      url = "https://raw.githubusercontent.com/void-linux/void-packages/4b97cd2fb4ec38712544438c2491b6d7d5ab334a/srcpkgs/sane/patches/sane-desc-cross.patch";
-      sha256 = "sha256-y6BOXnOJBSTqvRp6LwAucqaqv+OLLyhCS/tXfLpnAPI=";
-    })
-  ];
-
   postPatch = ''
+    # Do not create lock dir in install phase
+    sed -i '/^install-lockpath:/!b;n;c\       # pass' backend/Makefile.am
```
Comment (nh2) [inline pkgs/applications/graphics/sane/backends/default.nix:54]: Thank you for this improvement!

Could you explain shortly what exactly the `!b;n;c\` does?

Also, will this fail silently if upstream changes, and if yes is there a way to make it fail loudly, like e.g. https://nixos.org/manual/nixpkgs/stable/#fun-substitute-replace-fail ?

Thanks!

**Pull request merged**